File: javalib/src/main/scala/scala/scalanative/runtime/AtomicFieldUpdater.scala
package scala.scalanative.runtime

import java.util.concurrent.atomic.{
  AtomicIntegerFieldUpdater, AtomicLongFieldUpdater, AtomicReferenceFieldUpdater
}

import scala.scalanative.unsafe.Ptr

private class AtomicFieldUpdater {}
object AtomicFieldUpdater {
  def createIntegerFieldUpdater[T <: AnyRef](
      binding: AnyRef => Ptr[Int]
  ): AtomicIntegerFieldUpdater[T] =
    new AtomicIntegerFieldUpdater.Impl[T](binding)

  def createLongFieldUpdater[T <: AnyRef](
      binding: AnyRef => Ptr[Long]
  ): AtomicLongFieldUpdater[T] = new AtomicLongFieldUpdater.Impl[T](binding)

  def createReferenceFieldUpdater[T <: AnyRef, V <: AnyRef](
      binding: AnyRef => Ptr[V]
  ): AtomicReferenceFieldUpdater[T, V] =
    new AtomicReferenceFieldUpdater.Impl[T, V](binding)
}

File: nativelib/src/main/java/scala/scalanative/runtime/AtomicFieldUpdater.java
package scala.scalanative.runtime;

import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
import java.util.concurrent.atomic.AtomicLongFieldUpdater;
import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
import scala.Function1;

/** Compile-time facade for the Scala Native AtomicFieldUpdater object. */
public final class AtomicFieldUpdater {
  private AtomicFieldUpdater() {}

  public static <T> AtomicIntegerFieldUpdater<T> createIntegerFieldUpdater(Function1<Object, Object> binding) {
    throw new AssertionError("stub");
  }

  public static <T> AtomicLongFieldUpdater<T> createLongFieldUpdater(Function1<Object, Object> binding) {
    throw new AssertionError("stub");
  }

  public static <T, V> AtomicReferenceFieldUpdater<T, V> createReferenceFieldUpdater(Function1<Object, Object> binding) {
    throw new AssertionError("stub");
  }
}
